### PR TITLE
[optim] Allow torch.float64 scalars for forloop + foreach implementations

### PR DESCRIPTION
**Pull request by janeyx99**
Should allow for uses cases mentioned in #110940

This would allow scalars to also be float64s in the foreach implementation. The fused implementation would still create a float32 step on Adam and AdamW. This PR also does NOT worry about performance and is mainly for enablement.

Next steps:
- Relax the constraint on fused adam(w) and allow torch.float64 scalars there
- Allow _performant_ mixed dtypes in foreach (a bigger project in itself).

This PR will conflict with my other PRs, I will figure out a landing order

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115841

### CODE REVIEW

File: test/optim/test_optim.py [hunk 1]
```diff
@@ -770,11 +770,10 @@ def _test_derived_optimizers(self, optimizer_pairs_with_flags, flag, reduced_pre
...
                 # zero_size tensor as its last param.
                 # ref: https://github.com/pytorch/pytorch/issues/100701
-                empty_params = [torch.empty((), device=device, dtype=torch.float64)]
+                empty_param = torch.empty((), device=device, dtype=torch.float64)
+                parameters = list(model.parameters()) + [empty_param]
 
-                optimizer = optimizer_constructor(
-                    list(model.parameters()) + empty_params, **params_with_flags
-                )
+                optimizer = optimizer_constructor(parameters, **params_with_flags)
```
Comment (albanD) [inline test/optim/test_optim.py:777]: This is just refactoring right?

File: torch/optim/optimizer.py [hunk 2]
```diff
@@ -132,6 +132,11 @@ def _view_as_real(params, *state_and_grads):
             for s in state_and_grads:
                 s[i] = torch.view_as_real(s[i])
 
+def _get_scalar_dtype(is_fused=None):
+    if is_fused:
+        return torch.float32
+    return torch.float64 if torch.get_default_dtype() == torch.float64 else torch.float32
```
Comment (albanD) [inline torch/optim/optimizer.py:138]: I don't remember the story for default dtype vs default Tensor type. Do we need to check both here?

File: test/optim/test_optim.py (hunk 1 above)
Comment (janeyx99) [reply test/optim/test_optim.py:777]: This allows us to access the step of the first param later :p

File: torch/optim/optimizer.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/optimizer.py:138]: I don't think we do, get_default_dtype seems to work as expected:
image

**Review by albanD — APPROVED** [review]
Sounds good!

**Pull request closed without merging**